### PR TITLE
InferenceSession ctor with byte array in C#

### PR DESCRIPTION
**Pull request by yeohan**
**Description**: This is to add constructor overloads of InferenceSession object in C# to accept byte array in addition to file path. This allows us to pass a model image in memory, without having to serialize it into some temporary file.

**Motivation and Context**
I have a system that retains model images in memory. Without the new ctor overload, we'd have to serialize the image to some temporary file just to pass it to InferenceSession constructor. It is not a good idea because it causes extra file operation that has performance penalty and potential security risk.

With the fix, it will be straightforward to instantiate the InferenceSession object with a model image in memory. Say, just `var session = new InferenceSession(modelData)`.

### CODE REVIEW

**Comment by shahasad** [comment]
azp /run

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 1]
```diff
@@ -182,17 +201,30 @@ internal ModelMetadata ModelMetadata
...
         #region private methods
 
-        protected void Init(string modelPath, SessionOptions options)
+        /// <summary>
+        /// Initializes the session object with a model specified by either file path or byte array
+        /// </summary>
+        /// <param name="modelPath">Model's file path to load. This parameter is ignored if <paramref name="modelData"/> is specified.</param>
+        /// <param name="modelData">Model's byte array to read. If this parameter is ignored, then <paramref name="modelPath"/> will be ignored.</param>
+        /// <param name="options">Session options</param>
+        private void Init(string modelPath, byte[] modelData, SessionOptions options)
```
Comment (pranavsharma) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs]: Can this method be overloaded? Only one of modelPath and modelData can be present.

Comment (yeohan) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs]: Added two overloads accordingly.

**Comment by yeohan** [comment]
Will sync to latest master and resolve merge conflict on NativeMethods.cs.

**Comment by yeohan** [comment]
> Will sync to latest master and resolve merge conflict on NativeMethods.cs.

Done. Unit tests also passed.

I did rebase rather than merge to ensure to keep the baseline. Let me know if you see any issue wrt this.

**Pull request merged**